### PR TITLE
Restore simple text runs, correct for crashes

### LINKED ISSUE #6668: Reintroduce simple text analysis escape without crash or box drawing issue
Notes from bugs:

From #6488

> When I run this in PowerShell (7),
> 
> ```
> "`e[30;47m`u{2500} What `u{2500}`e[m"
> ```
> 
> the text comes out without antialiasing. It looks like it's being drawn with FillGeometry just like a box drawing glyph.
> 
> image
> 
> If I flip the colors, it comes out antialiased.
> 
> image

> Okay, this is because we're not populating `_glyphClusters` when text is "simple". When we go to split the run later (**we split runs for things other than text analysis!**), we try to remap the glyph clusters around the split.
> 
> Inevitably, we end up falling into DrawGlyphRun with `run.glyphCount` = 0 because `_SplitCurrentRun` stomped the glyphCount because there was no glyph cluster mapping.
> 
> Somehow _magically_ we end up doing a "get raw geometry, draw" for the entire string which very handily hides the glyph clustering bug!
> 
> This fix _works_ but I do not believe it is correct:
> 
> ```diff
> diff --git a/src/renderer/dx/CustomTextLayout.cpp b/src/renderer/dx/CustomTextLayout.cpp
> index d19d2d0c3..395a58224 100644
> --- a/src/renderer/dx/CustomTextLayout.cpp
> +++ b/src/renderer/dx/CustomTextLayout.cpp
> @@ -336,6 +336,7 @@ CustomTextLayout::CustomTextLayout(gsl::not_null<IDWriteFactory1*> const factory
>              for (size_t i = glyphStart; i < _glyphAdvances.size(); i++)
>              {
>                  _glyphAdvances.at(i) = (float)_glyphDesignUnitAdvances.at(i) / designUnitsPerEm * _format->GetFontSize() * run.fontScale;
> +                _glyphClusters.at(i) = static_cast<UINT16>(i);
>              }
> 
>              run.glyphCount = textLength;
> ```
> 
> I'm worried about that "magical" rendering fallthrough. With this fix, are we striking the text twice -- once as a geometry and once as a glyph run?
> 
> (I commented out the glyph run drawing, and I think the answer is _no_.)

From #6664

> Terminal crashed. In debug configuration raised exception in "vector" include file in
> function void _Verify_offset(const difference_type _Off) const at line 113
> **"cannot seek vector iterator after end"**
> 
> Call stack:
> ...
> problem at
> src\renderer\dx\CustomTextLayout.cpp at line 867:
> const auto postOriginX = std::accumulate(_glyphAdvances.begin() + run.glyphStart,
> _glyphAdvances.begin() + run.glyphStart + run.glyphCount,
> mutableOrigin.x);
> 
> _glyphAdvances.size(): 0x000000000000001f
> run.glyphStart: 0x00000001
> run.glyphCount: 0x0000ffff
> 
> Same problem in wsl2 (Ubuntu or CentOS not matter) with Midnight Commander... start ok (in user's home directory) but Terminal crashes when trying cd to /home directory with mouse double click under .. or with cd .. in command line.
> @igorkuzuro

### LINKED ISSUE #6669: Powerline glyphs hidden and some characters scaled improperly on re-rendering
/cc @miniksa

# Environment

```none
Windows build number: 10.0.20156.1000
Windows Terminal version: 1.2.1751.0
```

# Steps to reproduce

1. Open Terminal (my default profile is PowerShell 7 w/ Powerline theme, Cascadia Code PL font)
2. Open new tab (CMD)
3. Switch back to first tab
4. Switch back to second tab

# Expected behavior

Text looks consistent and all of it is present

# Actual behavior

Repro Gif

### PR DESCRIPTION
**Pull request by miniksa**
Restores the simple text run analysis and skipping of most of the
shaping/layout steps. Corrects one of the fast-path steps to ensure that
offsets and clusters are assigned.

## References
- Bug #6488 
- Bug #6664
- Simple run PR #6206 
- Simple run revert PR #6665
- Recycle glyph runs PR #6483

The "correction" functions, by which box drawing analysis is one of
them, is dependent on the steps coming before it properly assigning the
four main vectors of the text layout glyphs: indices, advances, offsets,
and clusters. When the fast path is identified by the code from #6206,
only two of those are fully updated: indices and advances. The offsets
doesn't tend to cause a problem because offsets are rarely used so
they're pretty much always 0 already (but this PR enforces that they're
zero for the simple/fast path.) The clusters, however, were not mapped
for the fast path. This showed itself in one of two ways:
1. Before the recycled runs PR #6483, the cluster map had a 0 in every
   field for the stock initialized vector.
2. After the recycled runs PR #6483, the cluster map had the previous
   run's mapping in it.

This meant that when we reached the steps where glyph runs were
potentially split during the correction phase for box drawing
characters, unexpected values were present to map the glyph indices to
clusters and were corrected, adjusted, or split in an unexpected
fashion. 

For instance, the index out of range bug could appear when the default 0
values appended to the end of the clusters vector were decremented down
to a negative value during the run splitter as the true DWrite cluster
mapper doesn't generate that sort of pattern in the slow path case
without also breaking the run itself.

The resolution here is therefore to ensure that all of the fields
related to glyph layout are populated even in the fast path. This
doesn't affect the slow path because that one always populated all
fields by asking DWrite to do it. The fast path just skips a bunch of
DWrite steps because it can implicitly identify patterns and save a
bunch of time.

I've also identified a few vectors that weren't cleared on reset/reuse
of the layout. I'm clearing those now so the `.resize()` operations
performed on them to get to the correct lengths will fill them with
fresh and empty values instead of hanging on to ones that may have been
from the previous. This should be OK memory perf wise because the act of
`.clear()` on a vector shouldn't free anything, just mark it invalid.
And doing `.resize()` from an empty one should just default construct
them into already allocated space (which ought to be super quick).

## Validation
* [x] Far.exe doesn't crash and looks fine
* [x] "\e[30;47m\u{2500} What \u{2500}\e[m" from #6488 appears
  appropriately antialiased
* [x] Validate the "\e[30;47m\u{2500} What \u{2500}\e[m" still works
  when `FillGeometry` is nerfed as a quick test that the runs are split
  correctly.
* [x] Put `u{fffd} into Powershell Core to make a replacement char in
  the output. Then press enter a few times and see that shrunken initial
  characters on random rows. Verify this is gone.

Closes #6668
Closes #6669

Co-Authored-By: Chester Liu <skyline75489@outlook.com>

### CODE REVIEW

**Comment by DHowett** [comment]
1. Gotta touch base with @carlos-zamora to see if this fixes #6669 
2. when committing, add a `Co-authored-by` from skyline

**Comment by DHowett** [comment]
Incremental diff from Chester's original code:

```diff
diff --git a/src/renderer/dx/CustomTextLayout.cpp b/src/renderer/dx/CustomTextLayout.cpp
index 4c78a74cf..50aa5eade 100644
--- a/src/renderer/dx/CustomTextLayout.cpp
+++ b/src/renderer/dx/CustomTextLayout.cpp
@@ -353,7 +353,6 @@ CATCH_RETURN()
             // With simple text, there's only one run. The actual glyph count is the same as textLength.
             _glyphDesignUnitAdvances.resize(textLength);
             _glyphAdvances.resize(textLength);
-            _glyphOffsets.resize(textLength);

             USHORT designUnitsPerEm = metrics.designUnitsPerEm;

@@ -368,6 +367,13 @@ CATCH_RETURN()
                 _glyphAdvances.at(i) = (float)_glyphDesignUnitAdvances.at(i) / designUnitsPerEm * _format->GetFontSize() * run.fontScale;
             }

+            // Empty all the offsets, we're not going to use them
+            _glyphOffsets.assign(textLength, {});
+
+            // Set all the clusters as sequential. In a simple run, we're going 1 to 1.
+            // Fill the clusters sequentially from 0 to N-1.
+            std::iota(_glyphClusters.begin(), _glyphClusters.end(), gsl::narrow_cast<unsigned short>(0));
+
             run.glyphCount = textLength;
             glyphStart += textLength;
```

File: src/renderer/dx/CustomTextLayout.cpp [hunk 1]
```diff
@@ -303,6 +341,45 @@ CATCH_RETURN()
...
+
+            // Empty all the offsets, we're not going to use them
+            _glyphOffsets.assign(textLength, {});
+
+            // Set all the clusters as sequential. In a simple run, we're going 1 to 1.
+            // Fill the clusters sequentially from 0 to N-1.
+            std::iota(_glyphClusters.begin(), _glyphClusters.end(), gsl::narrow_cast<unsigned short>(0));
+
+            run.glyphCount = textLength;
+            glyphStart += textLength;
```
Comment (DHowett) [inline src/renderer/dx/CustomTextLayout.cpp:378]: I'm a bit concerned about these three loops. One of them goes from `glyphStart` to `end`, and the other two touch _the entire collection_. Are the collections guaranteed to be the exact right length, and not have any values before `glyphStart`?

Regardless, is it possible to make these three loops just, like, one loop?

Comment (DHowett) [reply src/renderer/dx/CustomTextLayout.cpp:378]: (three loops = `for`, `assign`, `iota`)

**Comment by miniksa** [comment]
`Co-Authored-By: Chester Liu <skyline75489@outlook.com>`

**Comment by DHowett** [comment]
Actually, even reverting the original simplicity check didn't resolve #6669.

**Comment by miniksa** [comment]
Tests revalidated after `6443e54`

File: src/renderer/dx/CustomTextLayout.cpp (hunk 1 above)
Comment (miniksa) [reply src/renderer/dx/CustomTextLayout.cpp]: I cut the assign one as a part of other stuff. Now it's just two and I think the `iota` is elegant. So maybe leave it?

**Comment by skyline75489** [comment]
Overall it’s a solid PR. Thanks Michael!

Just some side note here: In #6488 ReLi suggested future work needs to be done to improve the simplicity analysis. I chatted with him and confirmed that we can trust GetTextComplexity for now. He also suggested that with potential complex scripting support in the future, the performance gain from simple text will, obvious, decrease. For now it does not really bothers me. I can sleep at night knowing that. We’ll see how it works out in the future.

**Comment by DHowett** [comment]
New incremental diff from Chester's OG code (for ease of review)

```diff
diff --git a/src/renderer/dx/CustomTextLayout.cpp b/src/renderer/dx/CustomTextLayout.cpp
index 4c78a74cf..7c14d6461 100644
--- a/src/renderer/dx/CustomTextLayout.cpp
+++ b/src/renderer/dx/CustomTextLayout.cpp
@@ -62,6 +62,12 @@ try
     _isEntireTextSimple = false;
     _textClusterColumns.clear();
     _text.clear();
+    _glyphScaleCorrections.clear();
+    _glyphClusters.clear();
+    _glyphIndices.clear();
+    _glyphDesignUnitAdvances.clear();
+    _glyphAdvances.clear();
+    _glyphOffsets.clear();
     return S_OK;
 }
 CATCH_RETURN()
@@ -212,8 +218,6 @@ CATCH_RETURN()
         // This result will be subdivided by the analysis processes.
         _runs.resize(1);
         auto& initialRun = _runs.front();
-        initialRun.nextRunIndex = 0;
-        initialRun.textStart = 0;
         initialRun.textLength = textLength;
         initialRun.bidiLevel = (_readingDirection == DWRITE_READING_DIRECTION_RIGHT_TO_LEFT);

@@ -353,7 +357,6 @@ CATCH_RETURN()
             // With simple text, there's only one run. The actual glyph count is the same as textLength.
             _glyphDesignUnitAdvances.resize(textLength);
             _glyphAdvances.resize(textLength);
-            _glyphOffsets.resize(textLength);

             USHORT designUnitsPerEm = metrics.designUnitsPerEm;

@@ -368,6 +371,10 @@ CATCH_RETURN()
                 _glyphAdvances.at(i) = (float)_glyphDesignUnitAdvances.at(i) / designUnitsPerEm * _format->GetFontSize() * run.fontScale;
             }

+            // Set all the clusters as sequential. In a simple run, we're going 1 to 1.
+            // Fill the clusters sequentially from 0 to N-1.
+            std::iota(_glyphClusters.begin(), _glyphClusters.end(), gsl::narrow_cast<unsigned short>(0));
+
             run.glyphCount = textLength;
             glyphStart += textLength;

@@ -590,6 +597,10 @@ CATCH_RETURN()
             //  1     1    .8 1  1
         }

+        // Dump the glyph scale corrections now that we're done with them.
+        _glyphScaleCorrections.clear();
+
+        // Order the runs.
         _OrderRuns();
     }
     CATCH_RETURN();
```

**Comment by miniksa** [comment]
AutoMerge back off so I can appropriately credit `Co-Authored-By: Chester Liu <skyline75489@outlook.com>`

**Comment by miniksa** [comment]
Dustin says I can put it in the commit text and it'll show up. So did that and turned automerge back on.

**Pull request merged**